File: src/java/Hillclimb.java
/*
 *    Ecotype Simulation models the sequence diversity within a bacterial
 *    clade as the evolutionary result of net ecotype formation and periodic
 *    selection, yielding a certain number of ecotypes.
 *
 *    Copyright (C) 2013-2014  Jason M. Wood, Montana State University
 *
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package ecosim;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.StringTokenizer;

/**
 *  Object to interact with the hillclimbing program.
 *
 *  @author Jason M. Wood
 *  @copyright GNU General Public License
 */
public class Hillclimb implements Runnable {

    /**
     *  Run the hillclimb program.
     *
     *  @param masterVariables The MasterVariables object.
     *  @param nu The number of environmental sequences.
     *  @param length The length of the sequences being analyzed.
     *  @param binning The Binning object.
     *  @param parameterSet The set of parameters to optimize.
     */
    public Hillclimb (MasterVariables masterVariables, Execs execs,
        Integer nu, Integer length, Binning binning,
        ParameterSet parameterSet) {
        this.masterVariables = masterVariables;
        this.execs = execs;
        this.nu = nu;
        this.length = length;
        this.binning = binning;
        this.parameterSet = parameterSet;
        String workingDirectory = masterVariables.getWorkingDirectory ();
        inputFileName = workingDirectory + "hillclimbIn.dat";
        outputFileName = workingDirectory + "hillclimbOut.dat";
        hasRun = false;
    }

    /**
     *  Run the hillclimb program.
     */
    public void run () {
        File inputFile = new File (inputFileName);
        File outputFile = new File (outputFileName);
        // Write the input values for the program to the hclimbIn.dat file.
        writeInputFile (inputFile);
        // Run the hillclimb program.
        execs.runHillclimb (inputFile, outputFile);
        // Get the output provided by the hillclimb program.
        result = readOutputFile (outputFile);
        // Set the flag stating that the hillclimb program has been run.
        if (result.getNpop () > 0) {
            hasRun = true;
        }
    }

    /**
     *  Returns true if hillclimb has been run, false otherwise.
     *
     *  @return True if hillclimb has been run, false otherwise.
     */
    public boolean hasRun () {
        return hasRun;
    }

    /**
     *  Returns the result of the hillclimb program.
     *
     *  @return The result.
     */
    public ParameterSet getResult () {
        return result;
    }

    /**
     *  Changes the value of hasRun.
     *
     *  @param hasRun The new value of hasRun.
     */
    public void setHasRun (boolean hasRun) {
        this.hasRun = hasRun;
    }

    /**
     *  Change the result stored in this object.
     *
     *  @param result The new result to store.
     */
    public void setResult (ParameterSet result) {
        this.result = result;
    }

    /**
     *  Returns the hillclimbing result as a String.
     *
     *  @return the hillclimbing result.
     */
    public String toString () {
        return result.toString ();
    }

    /**
     *  Private method to write the input file for the hillclimb program.
     *
     *  @param inputFile The file to write to.
     */
    private void writeInputFile (File inputFile) {
        ArrayList<BinLevel> bins = binning.getBins ();
        BufferedWriter writer = null;
        try {
            writer = new BufferedWriter (new FileWriter (inputFile));
            writer.write (String.format ("%-20d numcrit\n", bins.size ()));
            // Output the crit levels and the number of bins.
            for (int j = 0; j < bins.size (); j ++) {
                writer.write (String.format (
                    "%-20.6f %-20d\n",
                    bins.get (j).getCrit (),
                    bins.get (j).getLevel ()
                ));
            }
            // Write the omega value.
            writer.write (
                String.format ("%-20.5f omega\n", parameterSet.getOmega ())
            );
            // Write the sigma value.
            writer.write (
                String.format ("%-20.5f sigma\n", parameterSet.getSigma ())
            );
            // Write the npop value.
            writer.write (
                String.format ("%-20d npop\n", parameterSet.getNpop ())
            );
            // Write the nu value.
            writer.write (String.format ("%-20d nu\n", nu));
            // Write the nrep value.
            writer.write (String.format ("%-20d nrep\n", nrep));
            // Create the random number seed; an odd integer less than nine
            // digits long.
            long iii = (long)(100000000 * Math.random ());
            if (iii % 2 == 0) {
                iii ++;
            }
            // Write the random number seed.
            writer.write (
                String.format ("%-20d iii (random number seed)\n", iii)
            );
            // Write the length of the sequences.
            writer.write (
                String.format (
                    "%-20d lengthseq (after deleting gaps, etc.)\n",
                    length
                )
            );
            // Write the whichavg value.
            int whichavg = masterVariables.getCriterion ();
            writer.write (String.format ("%-20d whichavg\n", whichavg));

        }
        catch (IOException e) {
            System.out.println ("Error writing the input file.");
        }
        finally {
            if (writer != null) {
                try {
                    writer.close ();
                }
                catch (IOException e) {
                    System.out.println ("Error closing the input file.");
                }
            }
        }
    }

    /**
     *  Private method to read the output file from the hillclimb program.
     *
     *  @param outputFile The file to read from.
     */
    private ParameterSet readOutputFile (File outputFile) {
        ParameterSet result = new ParameterSet ();
        BufferedReader reader = null;
        try {
            reader = new BufferedReader (new FileReader (outputFile));
            String nextLine = reader.readLine ();
            while (nextLine != null) {
                StringTokenizer st = new StringTokenizer (nextLine);
                // There should only be one line containing omega, sigma,
                // npop, and the likelihood of that result.
                Double omega = new Double (st.nextToken ());
                Double sigma = new Double (st.nextToken ());
                Long npop = new Long (st.nextToken ());
                Double likelihood = new Double (st.nextToken ());
                result = new ParameterSet (npop, omega, sigma, likelihood);
                nextLine = reader.readLine ();
            }
        }
        catch (IOException e) {
            System.out.println ("Error reading the output file.");
        }
        finally {
            if (reader != null) {
                try {
                    reader.close ();
                }
                catch (IOException e) {
                    System.out.println ("Error closing the output file.");
                }
            }
        }
        return result;
    }

    private String inputFileName;
    private String outputFileName;

    private MasterVariables masterVariables;
    private Execs execs;
    private Integer nu;
    private Integer length;
    private Binning binning;
    private ParameterSet parameterSet;
    private ParameterSet result;

    private Integer nrep = 10000;

    private boolean hasRun;

}

File: src/fortran/hillclimb.f90
!    Ecotype Simulation models the sequence diversity within a bacterial clade as
!    the evolutionary result of net ecotype formation, periodic selection,
!    and drift, yielding a certain number of ecotypes.
! 
!    Copyright (C) 2009  Fred Cohan, Wesleyan University
!                        Danny Krizanc, Wesleyan University
!                        Jason M. Wood, Montana State University
!
!    This program is free software: you can redistribute it and/or modify
!    it under the terms of the GNU General Public License as published by
!    the Free Software Foundation, either version 2 of the License, or
!    (at your option) any later version.
!
!    This program is distributed in the hope that it will be useful,
!    but WITHOUT ANY WARRANTY; without even the implied warranty of
!    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!    GNU General Public License for more details.
!
!    You should have received a copy of the GNU General Public License
!    along with this program.  If not, see <http://www.gnu.org/licenses/>.


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Hill-climbing version of non-ultrametric design
! This is a non-ultrametric divergence design, with drift but no
! recombination. September 21, 2003
! note, this will soon be replaced by a quicker method that Danny
! is writing, which will not require the n-square binning algorithm,
! but rather an nlogn sorting routine
! note, the programs I've changed for nonultra are:
! main, whicheventandwhen, 
! divergenonultra (and deleted binning), binningdanny, coalescence,
! and startpops
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
program hillclimb
  use methods
  use simplexmethod
  implicit none
  character(len = 256)                :: inputFile
  character(len = 256)                :: outputFile
  logical                             :: fileExists
  integer                             :: i
  integer                             :: ier
  integer                             :: iprint
  integer                             :: iquad
  integer                             :: maxf
  integer                             :: nloop
  integer, parameter                  :: nparams = 4
  integer, parameter                  :: outputUnit = 3
  double precision                    :: simp
  double precision                    :: step(nparams)
  double precision                    :: stopcr
  double precision                    :: var(nparams)
  double precision                    :: params(nparams)
  double precision                    :: yvalue
  type(acinas_data)                   :: acinas
  type(number_increments)             :: numincs
  type(parameters_data)               :: bottom
  type(parameters_data)               :: top
  ! Nelder Mead function.
  procedure(nelmeadFunction), pointer :: functn
  functn => fredprogram
  ! Provide default file names to use.
  inputFile = 'hClimbIn.dat'
  outputFile = 'hClimbOut.dat'
  ! Read command line arguments.
  do i = 1, iargc()
    select case (i)
      case (1)
        call getArgument (1, inputFile)
      case (2)
        call getArgument (2, outputFile)
      case (3)
        call getArgument (3, numberThreads)
      case (4)
        call getArgument (4, debug)
      case default
        ! An unexpected number of arguments was supplied.
        write (unit = *, fmt = *) "Invalid number of paramenters supplied!"
        write (unit = *, fmt = *) "Expected: hClimbIn.dat hClimbOut.dat"
        ! Error, exit the program.
        stop
    end select
  end do
  ! Verify that the input file exists.
  inquire (file = trim (inputFile), exist = fileExists)
  if (fileExists .neqv. .true.) then
    write (unit = *, fmt = *) "The hclimbIn.dat file was not found at: ", &
      trim (inputFile)
    ! Error, exit the program.
    stop
  end if
  ! Read in the input file.
  call readinput (trim (inputFile), acinas, bottom, top, numincs)
  ! Open the output file.
  open (unit = outputUnit, file = trim (outputFile), access = 'sequential', form = 'formatted')
  ! Set max. no. of function evaluations = MAXF, print every IPRINT.
  maxf = 100
  iprint = 1
  ! Set value for stopping criterion.   Stopping occurs when the
  ! standard deviation of the values of the objective function at
  ! the points of the current simplex < stopcr.
  stopcr = 1.0d-1
  nloop = 8
  ! Fit a quadratic surface to be sure a minimum has been found.
  iquad = 0
  ! As function value is being evaluated in DOUBLE PRECISION, it
  ! should be accurate to about 15 decimals.   If we set simp = 1.d-6,
  ! we should get about 9 dec. digits accuracy in fitting the surface.
  simp = 1.0d-6
  params(1) = log (bottom%omega)
  params(2) = log (bottom%sigma)
  params(3) = bottom%npop
  params(4) = log (bottom%xn)
  step(1) = log (bottom%omega) / 2.0d0
  step(2) = log (bottom%sigma) / 2.0d0
  step(3) = bottom%npop / 2.0d0
  step(4) = log (bottom%xn) / 2.0d0
  yvalue = 0.0
  ! Danny, yvalue is just a dummy variable that gets passed to my
  ! program in the CALL statement, but the meaningful passing of 
  ! variables occurs through the COMMON block
  ! yvalue is the value returned for success level determined by jwhichxavg
  call nelmead (params, step, nparams, yvalue, maxf, iprint, stopcr, nloop, iquad, simp, var, functn, ier, &
    outputUnit, acinas%probthreshold)
  ! Close the random number generator.
  call randomClose ()
  ! Close the output file.
  close (unit = outputUnit)
  stop

  contains

  subroutine fredprogram (nparams, params, yvalue)
    ! Subroutine parameters
    integer, intent(in)             :: nparams
    double precision, intent(inout) :: params(nparams)
    double precision, intent(out)   :: yvalue
    ! Local variables
    real                  :: avgsuccess(6)
    type(parameters_data) :: parameters
    ! Make sure that npop is in the right range.
    if (params(3) .lt. 2) params(3) = 2
    if (params(3) .gt. acinas%nu) params(3) = acinas%nu
    parameters%omega = exp (params(1))
    parameters%sigma = exp (params(2))
    parameters%npop = nint (params(3))
    parameters%xn = exp (params(4))
    if (debug) then
      write (unit = *, fmt = *) 'omega= ', parameters%omega
      write (unit = *, fmt = *) 'sigma= ', parameters%sigma
      write (unit = *, fmt = *) 'npop= ', parameters%npop
      write (unit = *, fmt = *) 'xn= ', parameters%xn
    end if
    ! avgsuccess is a count of the number of results that are within X% tolerance
    ! for a particular set of parameter values.
    call simulation (acinas, parameters, avgsuccess)
    yvalue = -1.0d0 * avgsuccess(acinas%whichavg)
    if (debug) then
      write (unit = *, fmt = *) 'yvalue= ', yvalue
      write (unit = *, fmt = *)
    end if
    return
  end subroutine fredprogram

end program hillclimb
